contracts/DNSSECImpl.sol

```diff
@@ -387,153 +456,133 @@ contract DNSSECImpl is DNSSEC, Owned {
+            bytes memory keyrdata = proof.rdata();
+            RRUtils.DNSKEY memory dnskey = keyrdata.readDNSKEY(0, keyrdata.length);
+            if(verifySignatureWithKey(dnskey, keyrdata, rrset, data)) {
                 return true;
             }
         }
-
-        return false;
-    }
-
-    /**
-     * @dev Attempts to verify a signed RRSET against an already known public key.
-     * @param data The original data to verify.
-     * @param sig The signature data.
-     * @param offset The offset from the start of the data to the first RR.
-     * @return True if the RRSET could be verified, false otherwise.
-     */
-    function verifyWithDS(bytes memory data, bytes memory sig, uint offset, bytes memory proof) internal view returns(bool) {
-        // Extract algorithm and keytag
-        uint8 algorithm = data.readUint8(RRSIG_ALGORITHM);
-        uint16 keytag = data.readUint16(RRSIG_KEY_TAG);
-
-        // Perhaps it's self-signed and verified by a DS record?
-        for (RRUtils.RRIterator memory iter = data.iterateRRs(offset); !iter.done(); iter.next()) {
-            if (iter.dnstype != DNSTYPE_DNSKEY) {
-                return false;
-            }
-
-            bytes memory keyrdata = iter.rdata();
-            if (verifySignatureWithKey(keyrdata, algorithm, keytag, data, sig)) {
-                // It's self-signed - look for a DS record to verify it.
-                return verifyKeyWithDS(iter.name(), keyrdata, keytag, algorithm, proof);
-            }
-        }
-
-        return false;
     }
 
     /**
      * @dev Attempts to verify some data using a provided key and a signature.
-     * @param keyrdata The RDATA section of the key to use.
-     * @param algorithm The algorithm ID of the key and signature.
-     * @param keytag The keytag from the signature.
-     * @param data The data to verify.
-     * @param sig The signature to use.
+     * @param dnskey The dns key record to verify the signature with.
+     * @param rrset The signed RRSET being verified.
+     * @param data The original data `rrset` was decoded from.
      * @return True iff the key verifies the signature.
      */
-    function verifySignatureWithKey(bytes memory keyrdata, uint8 algorithm, uint16 keytag, bytes memory data, bytes memory sig)
+    function verifySignatureWithKey(RRUtils.DNSKEY memory dnskey, bytes memory keyrdata, RRUtils.SignedSet memory rrset, RRSetWithSignature memory data)
         internal
         view
         returns (bool)
     {
-        if (address(algorithms[algorithm]) == address(0)) {
-            return false;
-        }
         // TODO: Check key isn't expired, unless updating key itself
 
         // o The RRSIG RR's Signer's Name, Algorithm, and Key Tag fields MUST
         //   match the owner name, algorithm, and key tag for some DNSKEY RR in
         //   the zone's apex DNSKEY RRset.
-        if (keyrdata.readUint8(DNSKEY_PROTOCOL) != 3) {
+        if(dnskey.protocol != 3) {
```

Comment:
Should this be a constant rather than a magic number?  I know that https://tools.ietf.org/html/rfc4034#section-2.1.2 doesn't define options, so perhaps a reference to this as a comment would suffice?